applicationset/services/scm_provider/azure_devops_test.go

```diff
@@ -0,0 +1,287 @@
+
+			repo := &Repository{Organization: organization, Repository: repoName, RepositoryId: uuidHolder.uuid, Branch: branchName}
+			hasPath, err := provider.RepoHasPath(ctx, repo, path)
+
+			if testCase.returnError {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
+
+			assert.Equal(t, testCase.pathFound, hasPath)
+
+			gitClientMock.AssertExpectations(t)
+		})
+	}
+}
+
+func TestAzureDevopsGetBranches(t *testing.T) {
+	organization := "myorg"
+	teamProject := "myorg_project"
+	repoName := "myorg_project_repo"
+
+	ctx := context.TODO()
+	uuidHolder := getUuidHolder()
+
+	testCases := []struct {
+		name                string
+		expectedBranches    []azureGit.GitBranchStats
+		getBranchesApiError error
+		expectedError       error
+	}{
+		{
+			name:             "get_branches_single_branch_returns_branch",
+			expectedBranches: []azureGit.GitBranchStats{{Name: s("feature-feat1"), Commit: &azureGit.GitCommitRef{CommitId: s("abc123233223")}}},
+		},
+		{
+			name:                "get_branches_fails_return_empty_result",
+			getBranchesApiError: fmt.Errorf("Remote Azure Devops GetBranches error"),
+		},
+		{
+			name: "get_branches_no_branches_returned",
+		},
+		{
+			name:          "get_branches_get_client_fails",
+			expectedError: fmt.Errorf("Could not get Azure Devops API client"),
+		},
+		{
+			name: "get_branches_multiple_branches_returns_branches",
+			expectedBranches: []azureGit.GitBranchStats{
+				{Name: s("feature-feat1"), Commit: &azureGit.GitCommitRef{CommitId: s("abc123233223")}},
+				{Name: s("feature/feat2"), Commit: &azureGit.GitCommitRef{CommitId: s("4334")}},
+				{Name: s("feature/feat2"), Commit: &azureGit.GitCommitRef{CommitId: s("53863052ADF24229AB72154B4D83DAB7")}},
+			},
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			gitClientMock := azureMock.Client{}
+
+			clientFactoryMock := &AzureClientFactoryMock{mock: &mock.Mock{}}
+			clientFactoryMock.mock.On("GetClient", mock.Anything).Return(&gitClientMock, testCase.expectedError)
+
+			setup := gitClientMock.On("GetBranches", ctx, azureGit.GetBranchesArgs{RepositoryId: &repoName, Project: &teamProject}).Return(&testCase.expectedBranches, testCase.getBranchesApiError)
+			if testCase.expectedError != nil {
+				setup.Maybe()
+			}
+
+			repo := &Repository{Organization: organization, Repository: repoName, RepositoryId: uuidHolder.uuid}
+
+			provider := AzureDevopsProvider{organization: organization, teamProject: teamProject, clientFactory: clientFactoryMock}
+			branches, err := provider.GetBranches(ctx, repo)
+
+			if testCase.expectedError != nil {
+				assert.Error(t, err, "Expected error to occur on test case %s", testCase.name)
+			} else {
+				if testCase.getBranchesApiError != nil {
+					assert.Empty(t, branches)
+				} else {
+					if len(testCase.expectedBranches) > 0 {
+						assert.NotEmpty(t, branches)
+					}
+					assert.Equal(t, len(testCase.expectedBranches), len(branches))
```

Comment:
There's also an `assert.Len` if that sounds fun.  :-)

Reply:
It does :)